### DISCUSSION TITLE
Made a quick theme maker for customtkinter, make your own themes!!!

### DISCUSSION THREAD

**Discussion by Akascape** [TomSchimansky/CustomTkinter#264] [Showcase]
**NOW COMPATIBLE WITH THE LATEST CTK VERSION**
I made this quick theme creator for customtkinter with tkinter's default color chooser that can easily create themes for customtkiner
Try it: https://github.com/Akascape/CTkThemeMaker

(You can also quickly test the theme with a built in example)
## Current Version looks like this:
![Screenshot](https://github.com/Akascape/CTkThemeMaker/assets/89206401/69f91aa8-377e-4017-8a7d-9c7fb0ce110d)

**Comment by Valenterie** [comment]
That's dope ! Maybe the GUI is a little bit complicated and takes time to put all the colors but can't say that's not good, congrats !

**Reply by Akascape** [reply]
@Valenterie 
Thanks...
Its not that complicated if you study the json file properly, and I find it way better than the manual editing of json file where we have to put hex values of colors.

**Comment by TomSchimansky** [comment]
That's a good idea. But I think it would be better to have all widgets in one window, and then a dropdown menu at the bottom where you select a widget and then you select the colors for this widget. And the corresponding widget will automatically update to the new colors, so that you see how the colors match each other. Also it would be nice to have the option to load an existing theme file, if you just want to change some values.
But you can also make a PR for your current version, I will have a look at it.

**Reply by Akascape** [reply]
I made the PR for the current version...

**Comment by avalon60** [comment]
Superb! Is this downloadable anywhere?

**Reply by Akascape** [reply]
Yes, I have uploaded the thememaker program and also created a pull request for it.
https://github.com/Akascape/CustomTkinter/blob/master/examples/ThemeMaker.py

I don't know why Tom is not reviewing my pull request.

**Comment by avalon60** [comment]
Have you considered orienting this around a class? 

Also I would look at your naming conventions. Python is big on presentation, and there's  PEP which cover aspects of this (see later).

So for example, you have named a function (ChangeModeLeft) - this is using, camel case, the naming convention for a class. Functions should be  named in lower case, with underscores separating the words. I use Pycharm and it's also, by default, quite pedantic about spelling. So for example, it would complain if you named a function:

changemodeleft 

rather than:

change_mode_left

Also you might want to add doc string comments, as below, to each function, class etc:

E.g.

```
def transfer_config_section(source_file, target_file, section_name):
    """
    Transfer a section from a source config file to a target config file.

    :param source_file: The pathname to the source config file.
    :param target_file: The pathname to the target config file.
    :param section_name: The name of the section to copy.
    """
```
(When using  PyCharm,  it can  generate the basic template for the above format, using a keyboard shortcut)

So instead of describing a function (or class, or class method), by putting a hash comment before the  function, use a docstring. This allows tools like PyDoc etc, to expose the self-documenting aspects of your code. 

If you are not sharing your code, it may seem less important, but it needs to be presented well for a team to support it. Although I have never worked in an on-line community, I work in IT as a development manager. I don't know all of the "ins" and "outs" of how collaboration works in the on-line community, but working to common standards is important. You might want to take a look at the Python guidelines on naming conventions:

https://peps.python.org/pep-0008/

Nice stab at the tooling though!

**Reply by Akascape** [reply]
Thanks for your kind advice, I will definitely look into this. 
As I am a python learner from scratch, I never showed that much attention in the presentation part.

**Reply by avalon60** [reply]
@Akascape  If you haven't used classes before, you can get the basics from this Socratica video. I quite like these videos, because of the dead pan humour:

https://www.youtube.com/watch?v=apACNr7DC_s

At least I find them entertaining - maybe I'm a little odd... ;o)

**Reply by Akascape** [reply]
Thank you, I was just experimenting with the classes and somewhat successful in converting the whole program.

**Reply by Akascape** [reply]
@avalon60 Thanks for your kind guidance, I have successfully changed the code to class structure and also improved the code presentation. 
https://github.com/Akascape/CustomTkinter/blob/master/examples/ThemeMaker.py

**Comment by avalon60** [comment]
@Akascape , I have been having a play around with Theme Maker and I have a few suggestions, if I may:

1. It may be an idea to grab the list of "built in themes" from the list published in the ThemeManager class. Then use this list to block people from overwriting the standard themes. People breaking the standard themes, could cause Tom unnecessary headache.

2. The Dark / Light colour selection boxes have floating labels which, if the wrong colour is selected, makes the "Light" / "Dark" labels difficult to read. Maybe these could be external to the boxes OR have the label colours dynamically change to contrast with selected colour.

4. If a colour is selected, and you click the box again to adjust it, it doesn't persist the previously selected colour settings. Instead it resets the selections to 33 / 35 / 37.  It would be good, if it reflected the current values that were selected or loaded.

5. Using something like configparser and / or sqlite3 to remember settings / preferences / locations would be good. (see 6).

6. Theme Maker doesn't remember where you loaded a theme from. Consequently, when you save it, it the file isn't necessarily saved where you might expect.  You might have preferences which allow you to specify where themes are to be located or simply remember the last place used.

7.  When canceling the _Save Theme_ dialog, a spurious error message, "Something went wrong!" , is issued.

It would also be good if Tom were to enhance this snippet of code from theme_manager.py:

`        if theme_name_or_path in cls.built_in_themes:
            with open(os.path.join(script_directory, "assets", "themes", f"{theme_name_or_path}.json"), "r") as f:
                cls.theme = json.load(f)
        else:
            **with open(theme_name_or_path, "r") as f:
                cls.theme = json.load(f)**`

If the non standard theme reference doesn't include a json file extension, it would be nice to have the code in the _else_ block add one in.

Thanks,

Clive

**Comment by Akascape** [comment]
@avalon60 Good suggestions but I want to clear something,

1. Loading and editing the themes doesn't affect the main theme file until saved with the same name.
2. This is not a problem. There are only two labels which can't create confusion, still there is reset button for it. Moreover the text is still readable i think (don't know maybe my eyes work differently):

![Light](https://user-images.githubusercontent.com/89206401/180650997-cb5701d9-4282-4825-8ceb-38008057863a.png)  ![Dark](https://user-images.githubusercontent.com/89206401/180652722-47d85b37-ee9c-49dc-b6ce-331bfde04d9b.png)
It also shows the title in the askcolor widget.

3. In my case (windows) its not resetting the colors. Tkinter askcolor widget works different in other OSs. (Can't change it)
4. I don't think there is some special setting to be saved. Moreover I don't want to make the program so complex for the user.
5. I used the default CtkInputDialog and there is no special setting to change save location. But I just removed it with the default tkinter save dialog, now you can save the theme to a custom location (it also shows replace warning)
6. Its a fault of CtkInputDialog, I have opened the issue in #307

Thanks, :)

**Comment by avalon60** [comment]
1. "Loading and editing the themes doesn't affect the main theme file until saved with the same name."
       This is exactly the point. You are potentially allowing someone to overwrite Tom's JSON.

2. Not a problem for you ;o)

3. That must be blind luck. You are missing an option in the api. Example:  initialcolor=(33, 39, 55). There are several forms you can use. Please see: https://stackoverflow.com/questions/49221919/tkinter-default-colour-in-colorchooser

4. The setting could remember where the user last loaded the themes from. Currently I save a JSON file and then I  have to use an external file manager to copy / cut and paste it to the _venv/lib/python3.8/customtkinter/assets/themes_. Not only that, I have to navigate to that folder each time I run the program, so that I can load any JSON I want to modify. I then have to rinse and repeat the process. 
This can be solved by grabbing and saving the directory location. When the user opens the theme file, you can obtain  the directory pathname using the os module:
` self._json_dir = os.path.dirname(open_json)`
 This strips the filename from the end of the path and leaving only the path-name of the directory it was located.

5. Not sure why you would necessarily need a special setting in the widget. If you are using a library such as pathlib, you should be able to pr-append the load path to the json file name:    

```
self._json_dir = Path(self._json_dir)
theme_name =  self._json_dir / theme_name
```

6. This can be coded around:
```
    def save(self):
        dialog = customtkinter.CTkInputDialog(title="Save Theme", text="Enter your theme name:")
        try:
            # theme_name = dialog.get_input() + ".json"
            theme_name = dialog.get_input()
            if theme_name:
                theme_name = f"{theme_name}.json"
                with open(theme_name, "w") as f:
                    json.dump(self.json_data, f, indent=2)
                tkinter.messagebox.showinfo("Exported!", "Theme saved successfully!")
        except:
            tkinter.messagebox.showerror("Error!", "Something went wrong!")
```
Thanks,

Clive

**Reply by Akascape** [reply]
Fair enough...
Obviously have to improve the program in future. 

And as you mentioned that people can mess the original files, we can't really do much about it. Atleast the simple answer for them will be to reinstall customtkinter.

But for now, I think its good to go with the hacks.

Moreover, I think Tom will not add this program.

It will just remain as an external example created by me.

**Reply by avalon60** [reply]
@Akascape I have implemented most of this if you would like to take the code?

**Reply by TomSchimansky** [reply]
@Akascape I will have a look and review your code if I have the time maybe tomorrow or in the next days!

**Reply by Akascape** [reply]
> @Akascape I have implemented most of this if you would like to take the code?

@avalon60 Sure, you can coauthorised.

**Reply by avalon60** [reply]
Thanks @Akascape, the code is here. I have included a README.txt.

[theme_maker.py.zip](https://github.com/TomSchimansky/CustomTkinter/files/9178896/theme_maker.py.zip)

**Comment by Akascape** [comment]
@avalon60 Great changes, but these are my thoughts:

1) I think it will be good to create the `.theme_maker` folder in the same folder instead of __home_dir.  Secondly I have removed the CTkInputDialog with the tkinter save dialog, so now users can save the files to their custom locations. 
I think you have not checked my last committed thememaker.py.

Moreover dialogs automatically saves the last opened location in windows. I think it doesn't really matters to change the locations every time, maybe because I am used to it. Even other programs messes up these location dialogs.

2) I noticed that the program doesn't open when we directly run it (by double clicking).
3) The color picker doesn't remembers the color when we change the widget type. 

https://user-images.githubusercontent.com/89206401/180729330-f4a1836c-c7ca-41ae-a4ab-4acfe84d35f7.mp4

**Instead, initial color should be taken from the button.**

4) It shows error when we try to save the theme without loading.

So, overall its good with the other improvements.
And I will like to implement the other minor changes you made.

**Reply by avalon60** [reply]
@Akascape 

**"I think it will be good to create the .theme_maker folder in the same folder instead of __home_dir."** 
- On Linux, configurations are commonly located in hidden folders in the home directory. It's also a good anchor point. Think about what will happen if the program is invoked from a different location, via a full (or indirect) pathname. It won't be able to find the config folder, unless you add code to determine where the theme_maker.py lives, and then determine the config folder relative to that. 
Apologies, about the code version I based this on, as I say - I have not worked in an online community context before, so I am unsure of the process. In respect of the implementation of saving the last used location, that was meant as a minimal convenience feature. The other option is to have a preferences screen, where the location could be set and saved. This would require a preferences option and a new Toplevel window.


**"Moreover dialogs automatically saves the last opened location in _windows_. I think it doesn't really matters to change the locations every time, maybe because I am used to it. Even other programs messes up these location dialogs."**

- It's important to realise that when creating a tool such as this, it won't be used by only Windows users. Customtkinter is supported for MacOS, Linux (both similar) as well as Windows. The behaviour should be as consistent an experience as possible, irrespective of whether running on Windows or not. 
Apologies, about the code version I based this on, as I say - I have not worked in an online community context before, so I am unsure of the process. In respect of the implementation of saving the last used location, that was meant as a minimal convenience feature. The other option is to have a preferences screen, where the location could be set and saved. This would require a preferences option and a new Toplevel window.

**"I noticed that the program doesn't open when we directly run it (by double clicking)."** - That's possibly because Windows has no association with the file ".py" extension. I have never tried it on Windows, but on Linux, you can place a directive on the very first line. 
E.g. "#!/usr/bin/env python". This tells the O/S shell which interpreter to use.  I would be surprised if that method works on Windows, hence my thoughts on file associations.  If I get time, I'll have a play around on Windows - it's been a while since I developed anything primarily for windows. I think I might have achieved it with Perl, but that was years ago, and can't be sure. The other option, it to supply a BAT file which invokes Python with the theme_make.py script. Using wrapper scripts for Python, Java etc programs is one method sometimes used.

**"The color picker doesn't remembers the color when we change the widget type."** - yes I mentioned in the README.txt that this would be the next logical step. I think this would take a little more time and didn't do anything on that front. This would involve setting the self._color1 / self._colour2 class properties, each time the widget selection changes.

At the end of the day, these are my suggestions and this is your tool. I am happy to help you where I can, but the final technical/design decisions are yours.

**Reply by Akascape** [reply]
@avalon60 Its true that we should consider other OS, but still I think making/saving a config file makes the program a little more complex.

I also managed to fix the color picker by just adding `initialcolor=self.button_dark.fg_color`. No need to define `self._color1 = '#212325'`
It will just take the button color as initial that we have chosen. None color by default is #212325

Also added the disallowed message and improved my code (like `!=None` to `is not None`) 

Thanks for helping me, you are really awesome :)
